### PR TITLE
Add by row cleanup

### PR DESCRIPTION
**Pull request by ignaspa**
## Motivation and Context

If a csv file of annotation category information being uploaded has any issues, it can result in partially uploaded information which must then be cleaned up before reuploading a file without issues. This PR aims to remove any vestigial data from a partially successful upload.

## Your Changes

**Description**:
Adding a transaction to the execution of add_by_row.

**Type of change** (select all that apply):

 

- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have added tests for my changes. 
- [x] I have described any required documentation changes below. 

### Required documentation changes (if applicable)
Change wiki to explain that failed uploads will not result in partial data being kept.

### CODE REVIEW

File: app/models/annotation_category.rb [hunk 1]
```diff
@@ -23,62 +23,64 @@ class AnnotationCategory < ApplicationRecord
...
+                                              annotation_category: annotation_category.annotation_category_name)
+          end
         end
-        if new_deduction > criterion.max_mark || new_deduction < 0
-          raise CsvInvalidLineError, I18n.t('annotation_categories.upload.invalid_deduction',
-                                            annotation_content: text_with_deduction.first,
-                                            criterion_name: criterion_name,
-                                            value: new_deduction)
+      else
+        criterion = assignment.flexible_criteria.find_by(name: criterion_name)
```
Comment (ignaspa) [inline app/models/annotation_category.rb:54]: I think the STI for criteria is messing with this, but I can't remember if I am supposed to have the type or not now. @david-yz-liu

Comment (ignaspa) [reply app/models/annotation_category.rb:54]: (oh I rebased off master so you wouldn't have to deal with the merge conflict, so I guess I should check master?)

File: app/models/annotation_category.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@ def self.add_by_row(row, assignment, current_user)
         raise CsvInvalidLineError, I18n.t('annotation_categories.upload.empty_category_name')
       end
     elsif (annotation_category.flexible_criterion_id.nil? && !criterion_name.nil?) ||
-          (annotation_category.flexible_criterion.name != criterion_name)
+        (annotation_category.flexible_criterion.name != criterion_name)
```
Comment (david-yz-liu) [inline app/models/annotation_category.rb:36]: Please revert this change

File: spec/controllers/annotation_categories_controller_spec.rb [hunk 3]
```diff
@@ -475,7 +475,7 @@
 
       expect(response.status).to eq(302)
       # One annotation category was created, and one has an error.
```
Comment (david-yz-liu) [inline spec/controllers/annotation_categories_controller_spec.rb:477]: You should update this comment to make it clear that there is a valid annotation category in the file, but it isn't saved.
